In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("hw8.ipynb")

# CPSC 330 - Applied Machine Learning

## Homework 8: Introduction to Computer vision, Time Series, and Survival Analysis (Lectures 18 to 20)

## Imports

In [2]:
from hashlib import sha1

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer, make_column_transformer
from sklearn.pipeline import Pipeline, make_pipeline

from sklearn.preprocessing import StandardScaler, OrdinalEncoder, OneHotEncoder

from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor

from sklearn.metrics import r2_score

<div class="alert alert-info">
    
## Submission instructions
<hr>
rubric={points:2}

Follow the [homework submission instructions](https://github.com/UBC-CS/cpsc330-2024W2/blob/main/docs/homework_instructions.md). 

**You may work in a group on this homework and submit your assignment as a group.** Below are some instructions on working as a group.  
- The maximum group size is 2. 
- Use group work as an opportunity to collaborate and learn new things from each other. 
- Be respectful to each other and make sure you understand all the concepts in the assignment well. 
- It's your responsibility to make sure that the assignment is submitted by one of the group members before the deadline. 
- You can find the instructions on how to do group submission on Gradescope [here](https://help.gradescope.com/article/m5qz2xsnjy-student-add-group-members).


When you are ready to submit your assignment do the following:

1. Run all cells in your notebook to make sure there are no errors by doing `Kernel -> Restart Kernel and Clear All Outputs` and then `Run -> Run All Cells`. 
2. Notebooks with cell execution numbers out of order will have marks deducted. Notebooks without the output displayed may not be graded at all (because we need to see the output in order to grade your work).
3. Upload the assignment using Gradescope's drag and drop tool. Check out this [Gradescope Student Guide](https://lthub.ubc.ca/guides/gradescope-student-guide/) if you need help with Gradescope submission.
4. Make sure that the plots and output are rendered properly in your submitted file. 
5. If the .ipynb file is too big and doesn't render on Gradescope, also upload a pdf or html in addition to the .ipynb.

<br><br>

## Exercise 1: time series prediction

In this exercise we'll be looking at a [dataset of avocado prices](https://www.kaggle.com/neuromusic/avocado-prices). You should start by downloading the dataset and storing it under the `data` folder. We will be forcasting average avocado price for the next week. 

In [3]:
df = pd.read_csv("data/avocado.csv", parse_dates=["Date"], index_col=0)
df.head()

Date  AveragePrice  Total Volume     4046       4225    4770  \
0 2015-12-27          1.33      64236.62  1036.74   54454.85   48.16   
1 2015-12-20          1.35      54876.98   674.28   44638.81   58.33   
2 2015-12-13          0.93     118220.22   794.70  109149.67  130.50   
3 2015-12-06          1.08      78992.15  1132.00   71976.41   72.58   
4 2015-11-29          1.28      51039.60   941.48   43838.39   75.78   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region  
0     8696.87     8603.62       93.25          0.0  conventional  2015  Albany  
1     9505.56     9408.07       97.49          0.0  conventional  2015  Albany  
2     8145.35     8042.21      103.14          0.0  conventional  2015  Albany  
3     5811.16     5677.40      133.76          0.0  conventional  2015  Albany  
4     6183.95     5986.26      197.69          0.0  conventional  2015  Albany

In [4]:
df.shape

(18249, 13)

In [5]:
df["Date"].min()

Timestamp('2015-01-04 00:00:00')

In [6]:
df["Date"].max()

Timestamp('2018-03-25 00:00:00')

It looks like the data ranges from the start of 2015 to March 2018 (~2 years ago), for a total of 3.25 years or so. Let's split the data so that we have a 6 months of test data.

In [7]:
split_date = '20170925'
df_train = df[df["Date"] <= split_date]
df_test  = df[df["Date"] >  split_date]

In [8]:
assert len(df_train) + len(df_test) == len(df)

<br><br>

<!-- BEGIN QUESTION -->

### 1.1 How many time series? 
rubric={points:4}

In the [Rain in Australia](https://www.kaggle.com/datasets/jsphyg/weather-dataset-rattle-package) dataset from lecture demo, we had different measurements for each Location. 

We want you to consider this for the avocado prices dataset. For which categorical feature(s), if any, do we have separate measurements? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.1
    
</div>

_Points:_ 4

In [9]:
df.sort_values(by="Date").head()

Date  AveragePrice  Total Volume     4046      4225     4770  \
51 2015-01-04          1.75      27365.89  9307.34   3844.81   615.28   
51 2015-01-04          1.49      17723.17  1189.35  15628.27     0.00   
51 2015-01-04          1.68       2896.72   161.68    206.96     0.00   
51 2015-01-04          1.52      54956.80  3013.04  35456.88  1561.70   
51 2015-01-04          1.64       1505.12     1.27   1129.50     0.00   

    Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
51    13598.46    13061.10      537.36          0.0       organic  2015   
51      905.55      905.55        0.00          0.0       organic  2015   
51     2528.08     2528.08        0.00          0.0       organic  2015   
51    14925.18    11264.80     3660.38          0.0  conventional  2015   
51      374.35      186.67      187.68          0.0       organic  2015   

                region  
51           Southeast  
51             Chicago  
51  HarrisburgScranton  
51          Pittsburgh  
51               Boise

In [10]:
df[df['Date']=='2015-01-04'].shape

(108, 13)

From above, we can se that we have multiple rows for the same 'Date'. In fact, we have 108 rows of the same chosen date. We can clearly see that we obviously won't have different measurement for 'type' alone, since we also have multiple rows of the same 'Date' and 'type'. How about 'region'?

In [11]:
df.sort_values(by=["region", "Date"]).head()

Date  AveragePrice  Total Volume     4046      4225    4770  \
51 2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
51 2015-01-04          1.79       1373.95    57.42    153.88    0.00   
50 2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
50 2015-01-11          1.77       1182.56    39.00    305.12    0.00   
49 2015-01-18          1.17      44511.28   914.14  31540.32  135.77   

    Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
51     9716.46     9186.93      529.53          0.0  conventional  2015   
51     1162.65     1162.65        0.00          0.0       organic  2015   
50     8424.77     8036.04      388.73          0.0  conventional  2015   
50      838.44      838.44        0.00          0.0       organic  2015   
49    11921.05    11651.09      269.96          0.0  conventional  2015   

    region  
51  Albany  
51  Albany  
50  Albany  
50  Albany  
49  Albany

In [12]:
df['type'].unique()

array(['conventional', 'organic'], dtype=object)

We have two rows of the same 'Date' with the same 'region', but the 'type's are different (one row for 'conventional' and one for 'organic'). So each row of the dataset correspond to a different combination of 'type', 'region', and 'Date'. I can conclude that we should have separate measurement for each combination of 'region' and 'type'

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.2 Equally spaced measurements? 
rubric={points:4}

In the Rain in Australia dataset, the measurements were generally equally spaced but with some exceptions. How about with this dataset? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.2
    
</div>

_Points:_ 4

In [13]:
# modified code from lecture notes 19
def plot_time_spacing_distribution(df, region, type):
    """
    Plots the distribution of time spacing for a given region.
    
    Parameters:
        df (pd.DataFrame): The input DataFrame with columns 'Location' and 'Date'.
        region (str): The region (e.g., location) to analyze.
    """
    # Ensure 'Date' is in datetime format
    df['Date'] = pd.to_datetime(df['Date'])
    
    # Filter data for the given region
    region_data = df[(df['region'] == region) & (df['type'] == type)]
    
    if region_data.empty:
        print(f"No data available for region: {region}")
        return
    
    # Calculate time differences
    time_diffs = region_data['Date'].sort_values().diff().dropna()
    
    # Count the frequency of each time difference
    value_counts = time_diffs.value_counts().sort_index()
    
    # Display value counts
    string = (f"Time spacing counts for {region} and {type}:\n{value_counts}\n")
    return string, value_counts

In [14]:
string, value_counts = plot_time_spacing_distribution(df, 'Albany', 'organic')
print(string)

Time spacing counts for Albany and organic:
Date
7 days    168
Name: count, dtype: int64



In [15]:
print(plot_time_spacing_distribution(df, 'Albany', 'conventional')[0])

Time spacing counts for Albany and conventional:
Date
7 days    168
Name: count, dtype: int64



It seems like we have an equally spaced measurement with 7 days for each type of this particular region.

In [16]:
regions = df['region'].unique()
types = df['type'].unique()

for region in regions:
    for type in types:
        string, value_counts1 = plot_time_spacing_distribution(df, region, type)

        if not value_counts1.equals(value_counts):
            print(string)

Time spacing counts for WestTexNewMexico and organic:
Date
7 days     163
14 days      1
21 days      1
Name: count, dtype: int64



So all combination of type and region have an equal spacing of 7 days except for WestTexNewMexico and organic, which are mising on some weeks, thus have a 14 and 21 days spacing.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.3 Interpreting regions 
rubric={points:4}

In the Rain in Australia dataset, each location was a different place in Australia. For this dataset, look at the names of the regions. Do you think the regions are also all distinct, or are there overlapping regions? Justify your answer by referencing the data.

<div class="alert alert-warning">

Solution_1.3
    
</div>

_Points:_ 4

In [17]:
df['region'].unique()

array(['Albany', 'Atlanta', 'BaltimoreWashington', 'Boise', 'Boston',
       'BuffaloRochester', 'California', 'Charlotte', 'Chicago',
       'CincinnatiDayton', 'Columbus', 'DallasFtWorth', 'Denver',
       'Detroit', 'GrandRapids', 'GreatLakes', 'HarrisburgScranton',
       'HartfordSpringfield', 'Houston', 'Indianapolis', 'Jacksonville',
       'LasVegas', 'LosAngeles', 'Louisville', 'MiamiFtLauderdale',
       'Midsouth', 'Nashville', 'NewOrleansMobile', 'NewYork',
       'Northeast', 'NorthernNewEngland', 'Orlando', 'Philadelphia',
       'PhoenixTucson', 'Pittsburgh', 'Plains', 'Portland',
       'RaleighGreensboro', 'RichmondNorfolk', 'Roanoke', 'Sacramento',
       'SanDiego', 'SanFrancisco', 'Seattle', 'SouthCarolina',
       'SouthCentral', 'Southeast', 'Spokane', 'StLouis', 'Syracuse',
       'Tampa', 'TotalUS', 'West', 'WestTexNewMexico'], dtype=object)

Yes there is obviously an overlap. For example, we have Albany, Atlanta, and New York, which are cities in the New York State located in the north east. Yet we also have the broader Northeast region. Then, we also have TotalUS which combines all data accross the US. Distint regions would be one that only consist of cities, for example.

<!-- END QUESTION -->

<br><br>

We will use the entire dataset despite any location-based weirdness uncovered in the previous part.

We will be trying to forecast the avocado price. The function below is adapted from [Lecture 19](https://github.com/UBC-CS/cpsc330-2024W2/tree/main/lectures), with some improvements.

In [18]:
def create_lag_feature(df, orig_feature, lag, groupby, new_feature_name=None, clip=False):
    """
    Creates a new feature that's a lagged version of an existing one.
    
    NOTE: assumes df is already sorted by the time columns and has unique indices.
    
    Parameters
    ----------
    df : pandas.core.frame.DataFrame
        The dataset.
    orig_feature : str
        The column name of the feature we're copying
    lag : int
        The lag; negative lag means values from the past, positive lag means values from the future
    groupby : list
        Column(s) to group by in case df contains multiple time series
    new_feature_name : str
        Override the default name of the newly created column
    clip : bool
        If True, remove rows with a NaN values for the new feature
    
    Returns
    -------
    pandas.core.frame.DataFrame
        A new dataframe with the additional column added.
        
    """
        
    if new_feature_name is None:
        if lag < 0:
            new_feature_name = "%s_lag%d" % (orig_feature, -lag)
        else:
            new_feature_name = "%s_ahead%d" % (orig_feature, lag)
    
    new_df = df.assign(**{new_feature_name : np.nan})
    for name, group in new_df.groupby(groupby):        
        if lag < 0: # take values from the past
            new_df.loc[group.index[-lag:],new_feature_name] = group.iloc[:lag][orig_feature].values
        else:       # take values from the future
            new_df.loc[group.index[:-lag], new_feature_name] = group.iloc[lag:][orig_feature].values
            
    if clip:
        new_df = new_df.dropna(subset=[new_feature_name])
        
    return new_df

We first sort our dataframe properly:

In [19]:
df_sort = df.sort_values(by=["region", "type", "Date"]).reset_index(drop=True)
df_sort

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18244 2018-02-25          1.57      18421.24  1974.26   2482.65    0.00   
18245 2018-03-04          1.54      17393.30  1832.24   1905.57    0.00   
18246 2018-03-11          1.56      22128.42  2162.67   3194.25    8.93   
18247 2018-03-18          1.56      15896.38  2055.35   1499.55    0.00   
18248 2018-03-25          1.62      15303.40  2325.30   2171.66    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18244    13964.33    13698.27      266.06          0.0       organic  2018   
18245    13655.49    13401.93      253.56          0.0       organic  2018   
18246    16762.57    16510.32      252.25          0.0       organic  2018   
18247    12341.48    12114.81      226.67          0.0       organic  2018   
18248    10806.44    10569.80      236.64          0.0       organic  2018   

                 region  
0                Albany  
1                Albany  
2                Albany  
3                Albany  
4                Albany  
...                 ...  
18244  WestTexNewMexico  
18245  WestTexNewMexico  
18246  WestTexNewMexico  
18247  WestTexNewMexico  
18248  WestTexNewMexico  

[18249 rows x 13 columns]

We then call `create_lag_feature`. This creates a new column in the dataset `AveragePriceNextWeek`, which is the following week's `AveragePrice`. We have set `clip=True` which means it will remove rows where the target would be missing.

In [20]:
df_hastarget = create_lag_feature(df_sort, "AveragePrice", +1, ["region", "type"], "AveragePriceNextWeek", clip=True)
df_hastarget

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18243 2018-02-18          1.56      17597.12  1892.05   1928.36    0.00   
18244 2018-02-25          1.57      18421.24  1974.26   2482.65    0.00   
18245 2018-03-04          1.54      17393.30  1832.24   1905.57    0.00   
18246 2018-03-11          1.56      22128.42  2162.67   3194.25    8.93   
18247 2018-03-18          1.56      15896.38  2055.35   1499.55    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18243    13776.71    13553.53      223.18          0.0       organic  2018   
18244    13964.33    13698.27      266.06          0.0       organic  2018   
18245    13655.49    13401.93      253.56          0.0       organic  2018   
18246    16762.57    16510.32      252.25          0.0       organic  2018   
18247    12341.48    12114.81      226.67          0.0       organic  2018   

                 region  AveragePriceNextWeek  
0                Albany                  1.24  
1                Albany                  1.17  
2                Albany                  1.06  
3                Albany                  0.99  
4                Albany                  0.99  
...                 ...                   ...  
18243  WestTexNewMexico                  1.57  
18244  WestTexNewMexico                  1.54  
18245  WestTexNewMexico                  1.56  
18246  WestTexNewMexico                  1.56  
18247  WestTexNewMexico                  1.62  

[18141 rows x 14 columns]

Our goal is to predict `AveragePriceNextWeek`. 

Let's split the data:

In [21]:
df_train = df_hastarget[df_hastarget["Date"] <= split_date]
df_test  = df_hastarget[df_hastarget["Date"] >  split_date]

<br><br>

<!-- BEGIN QUESTION -->

### 1.4 `AveragePrice` baseline 
rubric={points}

Soon we will want to build some models to forecast the average avocado price a week in advance. Before we start with any ML though, let's try a baseline. Previously we used `DummyClassifier` or `DummyRegressor` as a baseline. This time, we'll do something else as a baseline: we'll assume the price stays the same from this week to next week. So, we'll set our prediction of "AveragePriceNextWeek" exactly equal to "AveragePrice", assuming no change. That is kind of like saying, "If it's raining today then I'm guessing it will be raining tomorrow". This simplistic approach will not get a great score but it's a good starting point for reference. If our model does worse that this, it must not be very good. 

Using this baseline approach, what $R^2$ do you get on the train and test data?

<div class="alert alert-warning">

Solution_1.4
    
</div>

_Points:_ 4

Snce we are always predicting this week's price, our predicted value is AveragePrice, while the actual value is AveragePriceNextWeek. So the $R^2$ score would be between AveragePriceNextWeek (y_true) and AveragePrice (y_pred)

In [22]:
from sklearn.metrics import r2_score
train_r2 = r2_score(df_train['AveragePriceNextWeek'], df_train['AveragePrice'])
train_r2

0.8285800937261841

In [23]:
test_r2 = r2_score(df_test['AveragePriceNextWeek'], df_test['AveragePrice'])
test_r2

0.7631780188583048

In [24]:
assert not train_r2 is None, "Are you using the correct variable name?"
assert not test_r2 is None, "Are you using the correct variable name?"
assert sha1(str(round(train_r2, 3)).encode('utf8')).hexdigest() == 'b1136fe2a8918904393ab6f40bfb3f38eac5fc39', "Your training score is not correct. Are you using the right features?"
assert sha1(str(round(test_r2, 3)).encode('utf8')).hexdigest() == 'cc24d9a9b567b491a56b42f7adc582f2eefa5907', "Your test score is not correct. Are you using the right features?"

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.5 Forecasting average avocado price
rubric={points:10}

Now that the baseline is done, let's build some models to forecast the average avocado price a week later. Experiment with a few approachs for encoding the date. Justify the decisions you make. Which approach worked best? Report your test score and briefly discuss your results.

Benchmark: you should be able to achieve $R^2$ of at least 0.79 on the test set. I got to 0.80, but not beyond that. Let me know if you do better!

Note: because we only have 2 splits here, we need to be a bit wary of overfitting on the test set. Try not to test on it a ridiculous number of times. If you are interested in some proper ways of dealing with this, see for example sklearn's [TimeSeriesSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html), which is like cross-validation for time series data.

<div class="alert alert-warning">

Solution_1.5
    
</div>

_Points:_ 10

In [25]:
df_train.head()

Date  AveragePrice  Total Volume     4046      4225    4770  \
0 2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1 2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2 2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3 2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4 2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0     9716.46     9186.93      529.53          0.0  conventional  2015   
1     8424.77     8036.04      388.73          0.0  conventional  2015   
2    11921.05    11651.09      269.96          0.0  conventional  2015   
3    10845.82    10103.35      742.47          0.0  conventional  2015   
4     9323.18     9170.82      152.36          0.0  conventional  2015   

   region  AveragePriceNextWeek  
0  Albany                  1.24  
1  Albany                  1.17  
2  Albany                  1.06  
3  Albany                  0.99  
4  Albany                  0.99

In [26]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15441 entries, 0 to 18222
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Date                  15441 non-null  datetime64[ns]
 1   AveragePrice          15441 non-null  float64       
 2   Total Volume          15441 non-null  float64       
 3   4046                  15441 non-null  float64       
 4   4225                  15441 non-null  float64       
 5   4770                  15441 non-null  float64       
 6   Total Bags            15441 non-null  float64       
 7   Small Bags            15441 non-null  float64       
 8   Large Bags            15441 non-null  float64       
 9   XLarge Bags           15441 non-null  float64       
 10  type                  15441 non-null  object        
 11  year                  15441 non-null  int64         
 12  region                15441 non-null  object        
 13  AveragePriceNextWeek 

It seems like we have no missing data

### Raw Ridge
Let's first try Ridge without encoding Date. Since we have a 'year' feature, we'll use that

In [27]:
numeric_features = [
    "AveragePrice",
    "Total Volume",
    "4046",
    "4225",
    "4770",
    "Total Bags",
    "Small Bags",
    "Large Bags",
    "XLarge Bags",
]
categorical_features = [
    "region",
    "type",
    "year"
]
drop_features = ["Date"]
target = ["AveragePriceNextWeek"]

In [28]:
# from lecture notes 19
def preprocess_features(
    df_train,
    df_test,
    numeric_features,
    categorical_features,
    drop_features,
    target
):

    all_features = set(numeric_features + categorical_features + drop_features + target)
    if set(df_train.columns) != all_features:
        print("Missing columns", set(df_train.columns) - all_features)
        print("Extra columns", all_features - set(df_train.columns))
        raise Exception("Columns do not match")

    numeric_transformer = make_pipeline(
        SimpleImputer(strategy="median"), StandardScaler()
    )
    categorical_transformer = make_pipeline(
        SimpleImputer(strategy="constant", fill_value="missing"),
        OneHotEncoder(handle_unknown="ignore", sparse_output=False),
    )

    preprocessor = make_column_transformer(
        (numeric_transformer, numeric_features),
        (categorical_transformer, categorical_features),
        ("drop", drop_features),
    )
    preprocessor.fit(df_train)
    ohe_feature_names = (
        preprocessor.named_transformers_["pipeline-2"]
        .named_steps["onehotencoder"]
        .get_feature_names_out(categorical_features)
        .tolist()
    )
    new_columns = numeric_features + ohe_feature_names

    X_train_enc = pd.DataFrame(
        preprocessor.transform(df_train), index=df_train.index, columns=new_columns
    )
    X_test_enc = pd.DataFrame(
        preprocessor.transform(df_test), index=df_test.index, columns=new_columns
    )

    y_train = df_train[target]
    y_test = df_test[target]

    return X_train_enc, y_train, X_test_enc, y_test, preprocessor

In [29]:
X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    df_train,
    df_test,
    numeric_features,
    categorical_features,
    drop_features, target
)

In [30]:
X_train_enc.head()

AveragePrice  Total Volume      4046      4225      4770  Total Bags  \
0     -0.432512     -0.234535 -0.229503 -0.222203 -0.214954   -0.232206   
1     -0.383676     -0.234440 -0.230948 -0.219448 -0.214272   -0.233587   
2     -0.554604     -0.233469 -0.231018 -0.219530 -0.214196   -0.229850   
3     -0.823205     -0.233283 -0.230996 -0.218170 -0.213945   -0.230999   
4     -0.994133     -0.225747 -0.230668 -0.196131 -0.213811   -0.232627   

   Small Bags  Large Bags  XLarge Bags  region_Albany  ...  region_Syracuse  \
0   -0.229907   -0.223154    -0.172063            1.0  ...              0.0   
1   -0.231513   -0.223789    -0.172063            1.0  ...              0.0   
2   -0.226469   -0.224325    -0.172063            1.0  ...              0.0   
3   -0.228629   -0.222193    -0.172063            1.0  ...              0.0   
4   -0.229930   -0.224856    -0.172063            1.0  ...              0.0   

   region_Tampa  region_TotalUS  region_West  region_WestTexNewMexico  \
0           0.0             0.0          0.0                      0.0   
1           0.0             0.0          0.0                      0.0   
2           0.0             0.0          0.0                      0.0   
3           0.0             0.0          0.0                      0.0   
4           0.0             0.0          0.0                      0.0   

   type_conventional  type_organic  year_2015  year_2016  year_2017  
0                1.0           0.0        1.0        0.0        0.0  
1                1.0           0.0        1.0        0.0        0.0  
2                1.0           0.0        1.0        0.0        0.0  
3                1.0           0.0        1.0        0.0        0.0  
4                1.0           0.0        1.0        0.0        0.0  

[5 rows x 68 columns]

In [31]:
X_train_enc.columns

Index(['AveragePrice', 'Total Volume', '4046', '4225', '4770', 'Total Bags',
       'Small Bags', 'Large Bags', 'XLarge Bags', 'region_Albany',
       'region_Atlanta', 'region_BaltimoreWashington', 'region_Boise',
       'region_Boston', 'region_BuffaloRochester', 'region_California',
       'region_Charlotte', 'region_Chicago', 'region_CincinnatiDayton',
       'region_Columbus', 'region_DallasFtWorth', 'region_Denver',
       'region_Detroit', 'region_GrandRapids', 'region_GreatLakes',
       'region_HarrisburgScranton', 'region_HartfordSpringfield',
       'region_Houston', 'region_Indianapolis', 'region_Jacksonville',
       'region_LasVegas', 'region_LosAngeles', 'region_Louisville',
       'region_MiamiFtLauderdale', 'region_Midsouth', 'region_Nashville',
       'region_NewOrleansMobile', 'region_NewYork', 'region_Northeast',
       'region_NorthernNewEngland', 'region_Orlando', 'region_Philadelphia',
       'region_PhoenixTucson', 'region_Pittsburgh', 'region_Plains',
       'r

In [32]:
def score_lr_print_coeff(preprocessor, train_df, y_train, test_df, y_test, X_train_enc):
    lr_pipe = make_pipeline(preprocessor, Ridge(random_state=123))
    lr_pipe.fit(train_df, y_train)
    print("Train score: {:.2f}".format(lr_pipe.score(train_df, y_train)))
    print("Test score: {:.2f}".format(lr_pipe.score(test_df, y_test)))
    lr_coef = pd.DataFrame(
        data=lr_pipe.named_steps["ridge"].coef_.flatten(),
        index=pd.DataFrame(X_train_enc).columns,
        columns=["Coef"],
    )
    return lr_coef.sort_values(by="Coef", ascending=False)

In [33]:
score_lr_print_coeff(preprocessor, df_train, y_train, df_test, y_test, X_train_enc)

Train score: 0.85
Test score: 0.80


Coef
AveragePrice                0.325180
region_SanFrancisco         0.091341
region_HartfordSpringfield  0.088819
region_NewYork              0.070054
type_organic                0.051681
...                              ...
region_Denver              -0.047878
type_conventional          -0.051681
region_SouthCentral        -0.067115
region_DallasFtWorth       -0.068354
region_Houston             -0.078587

[68 rows x 1 columns]

The result is not bad, and it is certainly better than our baseline. We can see from the coefficient that it relies heavily on AveragePrice. Let's try encoding the date.

## Encoding date as number (day since ...)

In [34]:
df_train = df_hastarget[df_hastarget["Date"] <= split_date]
df_test  = df_hastarget[df_hastarget["Date"] >  split_date]

In [35]:
first_day = df_train["Date"].min()

df_train = df_train.assign(
    Days_since=df_train["Date"].apply(lambda x: (x - first_day).days)
)
df_test = df_test.assign(
    Days_since=df_test["Date"].apply(lambda x: (x - first_day).days)
)
df_train.sort_values(by="Date").head()

Date  AveragePrice  Total Volume       4046       4225      4770  \
0     2015-01-04          1.22      40873.28    2819.50   28287.42     49.90   
13858 2015-01-04          0.94     461607.33  244152.26  165299.33  15302.75   
1352  2015-01-04          1.02     491738.00    7193.87  396752.18    128.82   
13689 2015-01-04          1.33       9213.49    3727.52    4327.52      0.00   
13520 2015-01-04          1.05     430138.88  110693.69  270107.61   9737.50   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
13858    36852.99    30884.29     5595.00        373.7  conventional  2015   
1352     87663.13    87406.84      256.29          0.0  conventional  2015   
13689     1158.45     1158.45        0.00          0.0       organic  2015   
13520    39600.08    39600.08        0.00          0.0  conventional  2015   

           region  AveragePriceNextWeek  Days_since  
0          Albany                  1.24           0  
13858    SanDiego                  0.82           0  
1352       Boston                  1.10           0  
13689  Sacramento                  1.27           0  
13520  Sacramento                  1.09           0

In [36]:
X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    df_train,
    df_test,
    numeric_features + ["Days_since"],
    categorical_features,
    drop_features,
    target
)
X_train_enc.columns

Index(['AveragePrice', 'Total Volume', '4046', '4225', '4770', 'Total Bags',
       'Small Bags', 'Large Bags', 'XLarge Bags', 'Days_since',
       'region_Albany', 'region_Atlanta', 'region_BaltimoreWashington',
       'region_Boise', 'region_Boston', 'region_BuffaloRochester',
       'region_California', 'region_Charlotte', 'region_Chicago',
       'region_CincinnatiDayton', 'region_Columbus', 'region_DallasFtWorth',
       'region_Denver', 'region_Detroit', 'region_GrandRapids',
       'region_GreatLakes', 'region_HarrisburgScranton',
       'region_HartfordSpringfield', 'region_Houston', 'region_Indianapolis',
       'region_Jacksonville', 'region_LasVegas', 'region_LosAngeles',
       'region_Louisville', 'region_MiamiFtLauderdale', 'region_Midsouth',
       'region_Nashville', 'region_NewOrleansMobile', 'region_NewYork',
       'region_Northeast', 'region_NorthernNewEngland', 'region_Orlando',
       'region_Philadelphia', 'region_PhoenixTucson', 'region_Pittsburgh',
       'regi

In [37]:
score_lr_print_coeff(preprocessor, df_train, y_train, df_test, y_test, X_train_enc)

Train score: 0.85
Test score: 0.75


Coef
AveragePrice                0.319366
region_SanFrancisco         0.097591
region_HartfordSpringfield  0.095290
region_NewYork              0.075457
region_Philadelphia         0.055506
...                              ...
region_Denver              -0.050518
type_conventional          -0.055376
region_SouthCentral        -0.072651
region_DallasFtWorth       -0.073299
region_Houston             -0.084002

[69 rows x 1 columns]

The results isn't that good. It is even worst than our baseline. This is because the encoded feature is seen as continuous. It doesn't capture any periodical patterns, etc. Let's try One Hot Encoding the month since Avocado prices could be affected by monthly/seasonal patterns.

## OHE of month

In [38]:
df_train = df_hastarget[df_hastarget["Date"] <= split_date]
df_test  = df_hastarget[df_hastarget["Date"] >  split_date]

In [39]:
df_train = df_train.assign(Month=df_train["Date"].apply(lambda x: x.month_name()))
df_test = df_test.assign(Month=df_test["Date"].apply(lambda x: x.month_name()))

In [40]:
df_train.head()

Date  AveragePrice  Total Volume     4046      4225    4770  \
0 2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1 2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2 2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3 2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4 2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0     9716.46     9186.93      529.53          0.0  conventional  2015   
1     8424.77     8036.04      388.73          0.0  conventional  2015   
2    11921.05    11651.09      269.96          0.0  conventional  2015   
3    10845.82    10103.35      742.47          0.0  conventional  2015   
4     9323.18     9170.82      152.36          0.0  conventional  2015   

   region  AveragePriceNextWeek     Month  
0  Albany                  1.24   January  
1  Albany                  1.17   January  
2  Albany                  1.06   January  
3  Albany                  0.99   January  
4  Albany                  0.99  February

In [41]:
X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    df_train, df_test, 
    numeric_features, 
    categorical_features + ["Month"], 
    drop_features,
    target
)
X_train_enc.columns

Index(['AveragePrice', 'Total Volume', '4046', '4225', '4770', 'Total Bags',
       'Small Bags', 'Large Bags', 'XLarge Bags', 'region_Albany',
       'region_Atlanta', 'region_BaltimoreWashington', 'region_Boise',
       'region_Boston', 'region_BuffaloRochester', 'region_California',
       'region_Charlotte', 'region_Chicago', 'region_CincinnatiDayton',
       'region_Columbus', 'region_DallasFtWorth', 'region_Denver',
       'region_Detroit', 'region_GrandRapids', 'region_GreatLakes',
       'region_HarrisburgScranton', 'region_HartfordSpringfield',
       'region_Houston', 'region_Indianapolis', 'region_Jacksonville',
       'region_LasVegas', 'region_LosAngeles', 'region_Louisville',
       'region_MiamiFtLauderdale', 'region_Midsouth', 'region_Nashville',
       'region_NewOrleansMobile', 'region_NewYork', 'region_Northeast',
       'region_NorthernNewEngland', 'region_Orlando', 'region_Philadelphia',
       'region_PhoenixTucson', 'region_Pittsburgh', 'region_Plains',
       'r

In [42]:
score_lr_print_coeff(preprocessor, df_train, y_train, df_test, y_test, X_train_enc)

Train score: 0.85
Test score: 0.80


Coef
AveragePrice                0.311424
region_SanFrancisco         0.106214
region_HartfordSpringfield  0.103903
region_NewYork              0.081879
region_Philadelphia         0.060428
...                              ...
region_Denver              -0.055380
type_conventional          -0.060229
region_SouthCentral        -0.077472
region_DallasFtWorth       -0.079459
region_Houston             -0.091224

[80 rows x 1 columns]

## OHE of season
The results from before is good, but let's try One Hot Encoding the season to capture a broader pattern.

In [43]:
# modified from lecture notes
def get_season(month):
    WINTER_MONTHS = ["December", "January", "February"] 
    AUTUMN_MONTHS = ["September", "October", "November"]
    SUMMER_MONTHS = ["June", "July", "August"]
    SPRING_MONTHS = ["March", "April", "May"]
    if month in WINTER_MONTHS:
        return "Winter"
    elif month in AUTUMN_MONTHS:
        return "Autumn"
    elif month in SUMMER_MONTHS:
        return "Summer"
    else:
        return "Fall"

In [44]:
df_train = df_train.assign(Season=df_train["Month"].apply(get_season))
df_test = df_test.assign(Season=df_test["Month"].apply(get_season))

In [45]:
df_train

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18218 2017-08-27          2.50      16137.93  2616.96   3672.96    0.00   
18219 2017-09-03          2.39       7657.47   927.27   4056.73    0.00   
18220 2017-09-10          2.38      11857.31  1562.10   4565.41    0.00   
18221 2017-09-17          2.36      10464.29  1845.14   2819.17    0.00   
18222 2017-09-24          2.26       9528.64  1545.34   2234.23    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18218     9848.01     9816.58       31.43          0.0       organic  2017   
18219     2673.47     2629.18       44.29          0.0       organic  2017   
18220     5729.80     5719.96        9.84          0.0       organic  2017   
18221     5799.98     5796.65        3.33          0.0       organic  2017   
18222     5749.07     5722.40       26.67          0.0       organic  2017   

                 region  AveragePriceNextWeek      Month  Season  
0                Albany                  1.24    January  Winter  
1                Albany                  1.17    January  Winter  
2                Albany                  1.06    January  Winter  
3                Albany                  0.99    January  Winter  
4                Albany                  0.99   February  Winter  
...                 ...                   ...        ...     ...  
18218  WestTexNewMexico                  2.39     August  Summer  
18219  WestTexNewMexico                  2.38  September  Autumn  
18220  WestTexNewMexico                  2.36  September  Autumn  
18221  WestTexNewMexico                  2.26  September  Autumn  
18222  WestTexNewMexico                  2.37  September  Autumn  

[15441 rows x 16 columns]

In [46]:
X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    df_train, df_test, 
    numeric_features, 
    categorical_features + ["Season"], 
    drop_features + ["Month"],
    target
)
X_train_enc.columns

Index(['AveragePrice', 'Total Volume', '4046', '4225', '4770', 'Total Bags',
       'Small Bags', 'Large Bags', 'XLarge Bags', 'region_Albany',
       'region_Atlanta', 'region_BaltimoreWashington', 'region_Boise',
       'region_Boston', 'region_BuffaloRochester', 'region_California',
       'region_Charlotte', 'region_Chicago', 'region_CincinnatiDayton',
       'region_Columbus', 'region_DallasFtWorth', 'region_Denver',
       'region_Detroit', 'region_GrandRapids', 'region_GreatLakes',
       'region_HarrisburgScranton', 'region_HartfordSpringfield',
       'region_Houston', 'region_Indianapolis', 'region_Jacksonville',
       'region_LasVegas', 'region_LosAngeles', 'region_Louisville',
       'region_MiamiFtLauderdale', 'region_Midsouth', 'region_Nashville',
       'region_NewOrleansMobile', 'region_NewYork', 'region_Northeast',
       'region_NorthernNewEngland', 'region_Orlando', 'region_Philadelphia',
       'region_PhoenixTucson', 'region_Pittsburgh', 'region_Plains',
       'r

In [47]:
score_lr_print_coeff(preprocessor, df_train, y_train, df_test, y_test, X_train_enc)

Train score: 0.85
Test score: 0.79


Coef
AveragePrice                0.313369
region_SanFrancisco         0.104028
region_HartfordSpringfield  0.101779
region_NewYork              0.080130
region_Philadelphia         0.059224
...                              ...
region_Denver              -0.054304
type_conventional          -0.059076
region_SouthCentral        -0.076319
region_DallasFtWorth       -0.077911
region_Houston             -0.089481

[72 rows x 1 columns]

The test score gets slightly lower. Maybe the prices have more meaningful variation between individual month than season. Let's add in 'Month' back and drop 'Season' since it is redundant.

In [48]:
df_train = df_train.assign(Month=df_train["Date"].apply(lambda x: x.month_name()))
df_test = df_test.assign(Month=df_test["Date"].apply(lambda x: x.month_name()))

In [49]:
X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    df_train, df_test, 
    numeric_features, 
    categorical_features + ["Month"], 
    drop_features + ["Season"],
    target
)
X_train_enc.columns

Index(['AveragePrice', 'Total Volume', '4046', '4225', '4770', 'Total Bags',
       'Small Bags', 'Large Bags', 'XLarge Bags', 'region_Albany',
       'region_Atlanta', 'region_BaltimoreWashington', 'region_Boise',
       'region_Boston', 'region_BuffaloRochester', 'region_California',
       'region_Charlotte', 'region_Chicago', 'region_CincinnatiDayton',
       'region_Columbus', 'region_DallasFtWorth', 'region_Denver',
       'region_Detroit', 'region_GrandRapids', 'region_GreatLakes',
       'region_HarrisburgScranton', 'region_HartfordSpringfield',
       'region_Houston', 'region_Indianapolis', 'region_Jacksonville',
       'region_LasVegas', 'region_LosAngeles', 'region_Louisville',
       'region_MiamiFtLauderdale', 'region_Midsouth', 'region_Nashville',
       'region_NewOrleansMobile', 'region_NewYork', 'region_Northeast',
       'region_NorthernNewEngland', 'region_Orlando', 'region_Philadelphia',
       'region_PhoenixTucson', 'region_Pittsburgh', 'region_Plains',
       'r

In [50]:
coeff_df = score_lr_print_coeff(preprocessor, df_train, y_train, df_test, y_test, X_train_enc)

Train score: 0.85
Test score: 0.80


In [51]:
coeff_df.loc[['Month_April',
       'Month_August', 'Month_December', 'Month_February', 'Month_January',
       'Month_July', 'Month_June', 'Month_March', 'Month_May',
       'Month_November', 'Month_October', 'Month_September']].sort_values(by='Coef', ascending=False)

Coef
Month_September  0.052539
Month_August     0.034566
Month_July       0.023132
Month_October    0.023126
Month_June       0.003469
Month_May        0.003266
Month_March     -0.002353
Month_November  -0.011705
Month_February  -0.015291
Month_April     -0.028658
Month_December  -0.033264
Month_January   -0.048827

We are now back to a test score of 0.80 by encoding month. We can see that the prices are the highest on September and lowest on January.

In all attempts, we have a same train score of 0.85. Encoding the date as number results in the lowest test score of 0.75, worse than our baseline. Encoding the season results in a better test score of 0.79. The seasons we encoded must have captured some patterns. The best results come from encoding the month with a test score of 0.80, just slightly higher than encoding seasons. This must be because months capture more patterns than the broader seasonal encoding. However, I noticed that this test score is the same as the very first attempt "Raw Ridge", where we didn't do further date encoding and relies only on the year feature. When looking at the final coefficients, we can see that on all attempts the prediction is highly influenced by AveragePrice, followed by regions. The coefficients of our encoded feature, as seen above, are quite low, meaning that our encoded date features doesn't affect the prediction much. So I think it make sense if encoding the date does not significantly improve performance.

<!-- END QUESTION -->

<br><br><br><br>

## Exercise 2: Short answer questions

<!-- BEGIN QUESTION -->

### 2.1 Time series

rubric={points:6}

The following questions pertain to Lecture 20 on time series data:

1. Sometimes a time series has missing time points or, worse, time points that are unequally spaced in general. Give an example of a real world situation where the time series data would have unequally spaced time points.
2. In class we discussed two approaches to using temporal information: encoding the date as one or more features, and creating lagged versions of features. Which of these (one/other/both/neither) two approaches would struggle with unequally spaced time points? Briefly justify your answer.
3. When studying time series modeling, we explored several ways to encode date information as a feature for the citibike dataset. When we used time of day as a numeric feature, the Ridge model was not able to capture the periodic pattern. Why? How did we tackle this problem? Briefly explain.

<div class="alert alert-warning">

Solution_2.1
    
</div>

_Points:_ 6

1. An example would be a time series recording the number of victims of a natural disaster, like earthquake, which does not happen on regular intervals.
2. The lagged version would struggle more because it depends on the value of previous point in time, and it assumes regular intervals. If the gap in this 'previous point in time' itself is not consistent, the lag column would not make sense and would be misleading. Encoding the date could work better, depending on how it is encoded. For example, encoding the day of the week or the month could still be meaningful because those features might still provide some information, even when the intervals of our data are irregular. 
3. This is because the numbers could not capture the periodic pattern. In a 3-hour interval, 00.00 comes after 21:00. When we encoded time as integer, Ridge cannot capture that 0 comes after 21. Instead, the linear model sees 0 and 21 as further apart. We tackled this by encoding the time of the day as a categorical feature, then added interaction features (eg. Mon 12:00) with PolynomialFeatures transformer to capture more patterns.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.2 Computer vision 
rubric={points:6}

The following questions pertain to Lecture 19 on multiclass classification and introduction to computer vision. 

1. How many parameters (coefficients and intercepts) will `sklearn`’s `LogisticRegression()` model learn for a four-class classification problem, assuming that you have 10 features? Briefly explain your answer.
2. In Lecture 19, we briefly discussed how neural networks are sort of like `sklearn`'s pipelines, in the sense that they involve multiple sequential transformations of the data, finally resulting in the prediction. Why was this property useful when it came to transfer learning?
3. Imagine that you have a small dataset with ~1000 images containing pictures and names of 50 different Computer Science faculty members from UBC. Your goal is to develop a reasonably accurate multi-class classification model for this task. Describe which model/technique you would use and briefly justify your choice in one to three sentences.

<div class="alert alert-warning">

Solution_2.2
    
</div>

_Points:_ 6

1. The number of coefficient will be $10*4=40$, one for each feature and class. The number of intercept would be 4, one for each class. So the total parameter would be 44.
2. Transfer learning is about reusing old models and fine tuning it. Since neural network apply transformation is layers, we could just reuse the early layers and fine tune some of the final steps. This would be easier and more efficient than training an entire model from scratch.
3. CNN (e.g. AlexNet) with transfer learning. Since we are dealing with images, it would not be a good idea to flatten them, and CNN can take in images without flattening them. However, training CNN from scratch is way too much work, and our dataset isn't that big, so it would be a great idea to fine tune it on our UBC CS faculty members dataset since it is already trained on famous datasets. We use it as feature extractor to get the feature vectors, then use classifier like Random Forest trained on our classes using the extracted feature.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

## 2.3 Survival analysis
<hr>

rubric={points:6}

The following questions pertain to Lecture 21 on survival analysis. We'll consider the use case of customer churn analysis.

1. What is the problem with simply labeling customers are "churned" or "not churned" and using standard supervised learning techniques?
2. Consider customer A who just joined last week vs. customer B who has been with the service for a year. Who do you expect will leave the service first: probably customer A, probably customer B, or we don't have enough information to answer? Briefly explain your answer. 
3. If a customer's survival function is almost flat during a certain period, how do we interpret that?

<div class="alert alert-warning">

Solution_2.3
    
</div>

_Points:_ 6

1. Whether a customer churn or not is time dependent. When we do our standard supervised learning technique, we predict whether a customer would churn or not in the time the data was collected, but customers behavior changes over time, and they can churn at any point of time. It would be more useful if we could determine when they are likely to churn.
2. I'd say we don't have enough information. Usually, like the example on the lecture, newer customers are expected to churn first. However, there are a lot of factors that affect how long a customer typically uses the service. It can depend on the type service, or the age of the customer, etc. Without knowing any info about the service or customer A and B, I can't say for sure.
3. The probability of them churning is stable on that period of time.

<!-- END QUESTION -->

<br><br>

**Before submitting your assignment, please make sure you have followed all the instructions in the Submission instructions section at the top.** 

![](img/eva-well-done.png)